src/components/SingleArticle.jsx

```diff
@@ -17,6 +21,19 @@ const SingleArticle = () => {
             });
     }, [article_id]);
 
+    const handleVote = (voteType) => {
+        const newVotesCount = voteType === 'up' ? article.votes + 1 : article.votes - 1;
+        setArticle({ ...article, votes: newVotesCount });
+        voteOnArticle(article_id, voteType)
+            .then((newArticle) => {
+                setArticle(newArticle);
+            })
```

Comment:
You do not need to update with articles again with the result from the database. 

You're doing optimistic rendering. Line 25/26 is updating the votes to look like what they should look like from the DB. You then just need to do the request, and if all goes well, the votes should look like what they are in the DB. 

However, if you update the votes, then do the request, but it doesnt succeed (i.e. no internet connection). The page now has updated votes, which dont match the databse - thats when you would need to reset them like you do in the catch.

At the moment, you get weird behaviour if you're on a slow connection, because its using the result from the db. 

- [ ] You need to remove this .then block to make the button work correctly on a slow connection, if I click the button multiple times at once.

Reply:
solved